velox/exec/tests/HashJoinTest.cpp

```diff
@@ -69,163 +95,327 @@ class HashJoinTest : public HiveConnectorTestBase {
+      for (const auto& file : files[i]) {
+        splits.push_back(exec::Split(makeHiveConnectorSplit(file->path)));
       }
+
+      splitInput.emplace(nodeIds[i], std::move(splits));
     }
+    return splitInput;
+  }
+
+  void runJoinTest(
+      const std::vector<std::string>& probeKeys,
+      std::vector<RowVectorPtr>& probeVectors,
+      const std::string& probeFilter,
+      const std::vector<std::string>& probeProjections,
+      bool isProbeParallelizable,
+      const std::vector<std::string>& buildKeys,
+      std::vector<RowVectorPtr>& buildBatches,
+      const std::string& buildFilter,
+      const std::vector<std::string>& buildProjections,
+      bool isBuildParallelizable,
+      const std::string& joinFilter,
+      const std::vector<std::string>& joinOutputLayout,
+      core::JoinType joinType,
+      const std::vector<std::string>& outputProjections,
+      const std::string& referenceQuery,
+      TestJoinVerifier testVerifier = nullptr) {
+    auto planNodeIdGenerator = std::make_shared<PlanNodeIdGenerator>();
+    auto planNode = PlanBuilder(planNodeIdGenerator)
+                        .values(probeVectors, isProbeParallelizable)
+                        .filter(probeFilter)
+                        .project(probeProjections)
+                        .hashJoin(
+                            probeKeys,
+                            buildKeys,
+                            PlanBuilder(planNodeIdGenerator)
+                                .values(buildBatches, isBuildParallelizable)
+                                .filter(buildFilter)
+                                .project(buildProjections)
+                                .planNode(),
+                            joinFilter,
+                            joinOutputLayout,
+                            joinType)
+                        .project(outputProjections)
+                        .planNode();
+    runJoinTest(std::move(planNode), {}, referenceQuery, testVerifier);
+  }
+
+  void runJoinTest(
+      core::PlanNodePtr&& planNode,
+      SplitInput&& inputSplits,
+      const std::string& referenceQuery,
+      TestJoinVerifier testVerifier = nullptr) {
+    CursorParameters params;
+    params.planNode = planNode;
+    params.maxDrivers = numDrivers_;
 
-    if (rightBatchSize != 0) {
-      opts.vectorSize = rightBatchSize;
-      fuzzer_.setOptions(opts);
-      for (int i = 0; i < numRightBatches; ++i) {
-        rightBatches.push_back(fuzzer_.fuzzRow(rightType));
+    {
+      SCOPED_TRACE("Without Spill");
+      bool noMoreSplits = false;
+      auto task = ::assertQuery(
+          params,
+          makeAddSplit(noMoreSplits, std::move(inputSplits)),
+          referenceQuery,
+          duckDbQueryRunner_);
+      // A quick sanity check for memory usage reporting. Check that peak total
+      // memory usage for the hash join node is > 0.
+      auto planStats = toPlanStats(task->taskStats());
+      auto joinNodeId = params.planNode->id();
```

Comment:
Isn't it the case that sometimes this node is not a join node, but a project node?

Reply:
Yes, I removed this check which doesn't seem to be very useful.